grafana/grafana
# Loki: Decouple from Prometheus operationUtils

Part of https://github.com/grafana/grafana/issues/72631

> [!NOTE]
>When it comes to Loki data source, it imports a lot of logic from `prometheus/querybuilder/shared/...`. We have discussed this between Prometheus & Loki data source teams and decided that:
> - for utils functions, we will duplicate them and have them in both - Loki and Prometheus. This is in long term beneficial, because as query languages (mostly in Loki) evolve, it will be easier to adjust logic and not change/adjust logic to work for both in another package that needs to be then released. 
> - components + related types will be moved to `@grafana/experimental` similarly as we do this with sql editor. We  also think that `@grafana/experimental` is a great place, as these components can be adjusted and in the future reused for other query builders, not just Prometheus and Loki. 
> 
> More in: https://docs.google.com/spreadsheets/d/1tRsIKOgyY5x-M-WxC8BGNUCDGrRWxOkzyABcPPBSauw/edit#gid=0

So this PR focuses on moving `operationUtils` functions that are needed in Loki to Loki. And moving `operationUtils` from `shared` as it is not going to be shared anymore.

Most of the PRs are just:
- changing the imports
- adding the needed code to Loki
- adding tests


To test:
1. Run `make devenv sources=loki,prometheus`
2. Turn on `Explain query` 
3. Build some queries in query builder and make sure are rendered correctly in preview and that explain is correctly rendered. Examples of queries:
- gdev-loki
  - `sum without(age) (rate({age="new", place="luna"} |= `` [$__auto])) + 2`
  - `count by(age) (rate({age="new", place="luna"} |= `` [$__auto])) / count by(age) (rate({age="new", place="luna"} |= `` | source = `data` [$__auto]))`
  - `sum(count_over_time({age="new"} |= `` | logfmt | __error__=`` [$__auto]))`
- gdev-prometheus 
  - `rate(grafana_http_request_duration_seconds_count{handler="/api/ds/query"}[$__rate_interval]) + 2` 
  - `sum without(job) (rate(grafana_http_request_duration_seconds_count{handler="/api/ds/query"}[$__rate_interval]))`
  - `histogram_quantile(0.95, sum by(le) (rate(grafana_http_request_duration_seconds_bucket{handler="/api/ds/query"}[$__rate_interval]))) / histogram_quantile(0.95, sum by(le, instance) (rate(grafana_http_request_duration_seconds_bucket[$__rate_interval])))`
 4. Make sure that in prometheus in Rate operation you see `$__rate_interval` and `$__interval` options and in Loki only `$__auto`
